store.go

```diff
@@ -219,14 +243,32 @@ func (sf *storeWithFallbacks) Get(ctx context.Context, serverAddress string) (au
 	return auth.EmptyCredential, nil
 }
 
-// Put saves credentials into the StoreWithFallbacks. It puts
-// the credentials into the primary store.
+// Put attempts to save the credentials into the primary store. If it encounters
+// ErrPlaintextPutDisabled, it will attempt to save the credentials into the
+// fallback stores one by one, until it succeeds or encounters other errors than
+// ErrPlaintextPutDisabled.
 func (sf *storeWithFallbacks) Put(ctx context.Context, serverAddress string, cred auth.Credential) error {
-	return sf.stores[0].Put(ctx, serverAddress, cred)
+	var err error
+	for _, s := range sf.stores {
+		err = s.Put(ctx, serverAddress, cred)
+		if err == nil {
+			return nil
+		}
+		if !errors.Is(err, ErrPlaintextPutDisabled) {
+			return err
+		}
+		// fallback to the next store on ErrPlaintextPutDisabled
+	}
+	return err
 }
 
-// Delete removes credentials from the StoreWithFallbacks for the given server.
-// It deletes the credentials from the primary store.
+// Delete removes the credentials from the primary store and all the fallback
+// stores.
 func (sf *storeWithFallbacks) Delete(ctx context.Context, serverAddress string) error {
```

Comment:
Or should we delete the first-found credential instead?

Reply:
Should we short circuit on first successful delete? I am worried that we might end up deleting creds from docker trust store when notation' truststore was different.

Reply:
@priteshbandi It's a bit tricky and here are my thoughts after some offline discussions:

There are two approaches to implement delete for `storeWithFallbacks`:
1. Delete the credentials in every store (The current implementation in this PR)
2. Delete the first-found credentials and short circuit

Let's take notation as an example and look at the following scenarios:

### Scenario A

#### notation's creds store: pass
`{ "example.com": valid_cred1 }`
#### docker's creds store: desktop.exe
`{ "example.com": valid_cred2 }`

- Approach 1: After `notation logout example.com`, `valid_cred1` and `valid_cred2` are both deleted. commands like `notation sign` and `docker pull` will fail.
- Approach 2: After `notation logout example.com`, `valid_cred1` is deleted while `valid_cred2` is kept. `notation sign` will still succeed as notation can read `valid_cred2` from the docker's creds store; `docker pull` will succeed.

### Scenario B
#### notation's creds store: pass
`{ "example.com": expired_cred1 }`
#### docker's creds store: desktop.exe
`{ "example.com": valid_cred2 }`

Before `notation logout example.com`, `notation sign` would fail as it would use `expired_cred1` which would not work, and `docker pull` would succeed.
- Approach 1: After `notation logout example.com`, `expired_cred1` and `valid_cred2` are both deleted. commands like `notation sign` and `docker pull` will fail.
- Approach 2:  After `notation logout example.com`, `expired_cred1` is deleted while `valid_cred2` is kept. **Interestingly, `notation sign` will succeed as notation can read valid_cred2 from the docker's creds store;** `docker pull` will succeed.

### Scenario C
#### notation's creds store: pass
`{ "example.com": valid_cred }`
#### docker's creds store: pass
`{ "example.com": valid_cred }`

Before `notation logout example.com`, commands like `notation sign` and `docker pull` would succeed.
- Approach 1: After `notation logout example.com`, `valid_cred` is deleted from the shared creds store. `notation sign` and `docker pull` will fail.
- Approach 2: After `notation logout example.com`, `valid_cred` is deleted from the shared creds store. `notation sign` and `docker pull` will fail.


There are acutally more scenarios. But we can see that the approach 1 is deterministic as it always delete the docker's credentials, and `notation logout` will always fail the subsequent `docker pull`. While approach 2 may or may not delete docker's credentials, that means `notation logout` will sometimes fail `notation sign` and `docker pull`, sometimes not, depending on the configuration.

IMHO it's better to have a deterministic behavior so that users can easily expect the result of the `logout` command. 
@shizhMSFT @yizha1 may add some comments.

Reply:
It turns out that either approach is not ideal. Updated per the suggestions at https://github.com/notaryproject/notation/pull/712#issuecomment-1584042555.